### DISCUSSION TITLE
@mantine/form: Expose FormValidateInput or validation in external file

### DISCUSSION THREAD

**Discussion by guy-ecton** [mantinedev/mantine#6189] [Q&A] [answered]
I'm using `@mantine/form` and want to write the validation rule in a file.
I get the list of fields from the DB so I don't know their names. But I do know their types.

I want to write a function that dynamically creates the `validation` object, and need access to it `FormValidateInput`.

```typescript
// Comp.tsx
import validationRules from './validation.ts'

function Comp() {
  const fields: Field[] = ...
  const form = useForm({
   ...
   validation: validationRules(fields)
  })
}

// validation.ts
function createRules(fields: Field[]) {
  return fields.reduce((acc, field) => {
    if (field.type === 'boolean') {
     acc[field.name] = (value: boolean) => typeof value === "boolean" ? null : 'invalid type'
    } else {
     acc[field.name] = isNotEmpty()
    }
   return acc
  },
 {} // <-- typescript needs a type in here
 )
}
```

Thanks

**Comment by rtivital** [comment]
I do not understand what you need from the `@mantine/form` library

**Reply by guy-ecton** [reply]
I need to supply proper typing for that `{}` (initial value in reduce). Or else, typescript yells at me.

**Reply by rtivital** [reply]
If you know exact type, then specify it. If you don't know it, use Record<string, any>

**Reply by guy-ecton** [reply]
So, the exact type, as I understand it, is `FormValidateInput`.
Having `Record<string, any>` is not narrow enough, which impacts DX for us.

**Reply by rtivital** [reply, marked as answer]
If you really need this, use `UseFormInput` type from `@mantine/form` package. To learn how to use it, see https://github.com/mantinedev/mantine/blob/master/packages/%40mantine/form/src/types.ts
